staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go

```diff
@@ -196,6 +196,18 @@ type ValidationRule struct {
 	// If unset, the message is "failed rule: {Rule}".
 	// e.g. "must be a URL with the host matching spec.host"
 	Message string
+	// MessageExpression declares a CEL expression that evaluates to the validation failure message that is returned when this rule fails.
+	// Since messageExpression is used as a failure message, it must evaluate to a string.
+	// If both message and messageExpression are present on a rule, then messageExpression will be used if validation
+	// fails. If messageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
+	// as if the messageExpression field were unset. If messageExpression evaluates to an empty string, a string with only spaces, or a string
+	// that contains line breaks, then the validation failure message will also be produced as if the messageExpression field were unset, and
+	// the fact that messageExpression produced an empty string/string with only spaces/string with line breaks will be logged.
+	// messageExpression has access to all the same variables as the rule; the only difference is the return type.
```

Comment:
@jpbetz - did the env scoping get put in place so CRD CEL validation didn't automatically get access to the authorizer variable? is there anything that verifies validation rules and messageExpressions don't have that variable?

Reply:
Yes. CRD validation rules go different path from ValidatingAdmissionPolicy hence it will not have access to authorizer related variables

Reply:
We will make sure messageExpression not allowed to access to authorizer in here: https://github.com/kubernetes/kubernetes/pull/116397

Reply:
> Joe Betz - did the env scoping get put in place so CRD CEL validation didn't automatically get access to the authorizer variable? 

Yes, that's all in place.

> is there anything that verifies validation rules and messageExpressions don't have that variable?

Yes, when the secondary authz PR merged it included this test: https://github.com/kubernetes/kubernetes/blob/90c3232de77aa0dd09b948ffdd27c575688fba8a/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go#L1762